### PR TITLE
Meson build helper: make sure env variables from conan profile take precedence

### LINKED ISSUE #5999: [bug] Meson build helper overrides profile env variables when compiling with MSVC+ninja
When building with Conan+Meson+ninja on Windows, conan ends up shadowing PATH env variable's value specified in the corresponding conan profile with the value coming from "vcvars". For example, given this profile:

```
SQUISHCOCO="C:\Program Files\squishcoco"

[env]
PATH=[$SQUISHCOCO\visualstudio_x64;$SQUISHCOCO]
COVERAGESCANNER_ARGS=--cs-on --cs-mcdc --cs-mcc
```

one would expect the PATH variable at the time of the Meson invocation to start with `C:\Program Files\squishcoco\visualstudio_x64;C:\Program Files\squishcoco;`. In practice, it ends up looking more like this: `C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.15.26726\bin\HostX64\x64;<more MSVC paths...>;C:\Program Files\squishcoco\visualstudio_x64;C:\Program Files\squishcoco;<more paths...>`

**Some context**: [Squish Coco](https://doc.froglogic.com/squish-coco/latest/) is a commercial, cross-platform C/C++ code coverage tool often used in heavily-regulated software industries (medical devices, avionics, etc.). To instrument your code, you have to configure your `PATH` to point to Squish Coco's compiler/librarian/linker wrappers that mimic/redirect to the native toolchain. This issue prevents us from being able to use Conan when compiling our projects under Squish Coco.

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version: Visual Studio 2017
  * Conan version: 1.20.0-dev
  * Python version: 3.8.0

### PR DESCRIPTION
**Pull request by agurtovoy**
Changelog: Bugfix: fixes #5999 (with a test)
Docs: omit

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by memsharded** [comment]
Hi @agurtovoy 

Thanks for your contribution. I have reviewed it and I think we need first to understand the scope of this issue, and if it affects other build systems. From what I am seeing in the code, it seems that ``CMake`` is not automatically adding the environment as your PR does. And we need to keep the behavior of build helpers consistent.

**Comment by agurtovoy** [comment]
@memsharded Good point. I should note that, in general, the profile env vars are already properly set by the time a build helper gets control; the issue is that some build helpers invoke `tools.vcvars` after that, effectively invalidating any prior env setup, specifically as it's concerned with `PATH`.

**Comment by lasote** [comment]
I understand the issue but we cannot solve it applying the `conanfile.env` at that point. As Diego said, all the build helpers act the same way. We might think in another mechanism to skip/force something?

**Comment by agurtovoy** [comment]
@lasote I totally understand the desire to keep the environment setup in the common code, but it can be argued that build helpers already violate that rule/make it moot by applying their own changes to the environment — after the common setup has been run — on a conditional, ad hoc basis. It's not immediately apparent to me why the latter is okay, but making it the helper's responsibility to insure that the user-specified environment survives that fiddling is not.

**Comment by memsharded** [comment]
Hi @agurtovoy 

I have been playing with this here: https://github.com/conan-io/conan/pull/6297
Considering to merge it, but I am still afraid that this could break others users that actually want this behavior and rely on the helper for defining the right environment. I wanted to ask. Would you be ok with an opt-in mechanism, something like:

```python
cmake = CMake(..., apply_self_env=True)
cmake.build()
```

This way we can introduce this functionality being 100% sure we don't break other users.

**Comment by agurtovoy** [comment]
Hi @memsharded, we would be okay with since we're always constructing build helpers manually, and I totally understand the desire to ensure backward compatibility, but it seems to me that the fix has the possibility of breaking something only if the user is using Conan's `[env]` section to override/add environment variables, and if they are, then, in accordance to the current docs, they expect these overrides to work. Having to explicitly opt-in to what is currently documented as default behavior seems off to me. 

Hope this makes sense, and thanks for all your work.

**Comment by jgsogo** [comment]
Hi, although I approved #6297, after thinking about this issue and sharing my POV with some other team members I've changed my opinion and, thinking about the long term, I think that the environment from the profile cannot modify the behavior of our build helpers. 

(Sorry if it sounds a little bit opinionated, I'm open to talk about it, of course, but I want to make sure it is clear) 👨‍👧‍👧  

IMHO a **Build helper is the implementation of a commitment between the Conan settings and the build system**, same inputs should provide the same outputs and we should do our best to keep this rule of thumb. Environment variables are a hidden way to modify the build helper behavior, if we let the user change some behavior (choosing a different compiler, for example) we are like hacking the build helper.

I don't like **hidden** environment variables, but I would support having a way to pass some attributes (even environment variables) like the one that @memsharded is pointing out, nevertheless I would force it to be even more explicit:

```python
cmake = CMake(...)
...
env = {... build your dictionary here... }
cmake.build(env=env)
```

and this `env` can be applied just before running the actual command (after applying the env from the dependencies and the env from the build-helper itself).

I think that any move we do from now on should take into account the behavior we want for next Conan major release and **without breaking** we should solve the bugs and implement the new features having that future release in mind.

@agurtovoy I need to have a look at the docs and see if this is too much breaking according to them. Maybe if we agree on this behavior the _bug_ is on the docs and we can _fix them_ to explain better the expected behavior (and the intention for the next big release)

File: conans/client/tools/env.py [hunk 1]
```diff
@@ -30,13 +30,14 @@ def run_environment(conanfile):
 
 
 @contextmanager
-def environment_append(env_vars):
+def environment_append(env_vars, post=False):
```
Comment (jgsogo) [inline conans/client/tools/env.py]: Do we want to add it to the tool and to expose the new behavior to the user?  (no one has asked for it so far)

File: conans/client/build/meson.py [hunk 2]
```diff
@@ -159,15 +162,22 @@ def configure(self, args=None, defs=None, source_dir=None, build_dir=None,
...
-        with tools.vcvars(self._settings,
-                          output=self._conanfile.output) if self._vcvars_needed else tools.no_op():
+        def _build():
             env_build = AutoToolsBuildEnvironment(self._conanfile)
             with environment_append(env_build.vars):
                 self._conanfile.run(command)
 
+        if self._vcvars_needed:
+            vcvars_dict = tools.vcvars_dict(self._settings, output=self._conanfile.output)
+            with environment_append(vcvars_dict, post=self._append_vcvars):
```
Comment (jgsogo) [inline conans/client/build/meson.py]: Maybe the `post=self._append_vcvars` makes sense as an argument to `with tools.vcvars(..., post=self._append_vcvars)` and keep `environment_append` unmodified.

File: conans/client/build/cmake.py [hunk 3]
```diff
@@ -214,10 +215,11 @@ def _run(self, command):
...
         is_msvc = compiler == "Visual Studio"
-        if (is_msvc or is_clangcl) and self.generator in ["Ninja", "NMake Makefiles",
-                                                          "NMake Makefiles JOM"]:
-            with tools.vcvars(self._settings, force=True, filter_known_paths=False,
-                              output=self._conanfile.output):
+        if ((is_msvc or is_clangcl) and platform.system() == "Windows" and
+                self.generator in ["Ninja", "NMake Makefiles", "NMake Makefiles JOM"]):
+            vcvars_dict = tools.vcvars_dict(self._settings, force=True, filter_known_paths=False,
+                                            output=self._conanfile.output)
+            with environment_append(vcvars_dict, post=self._append_vcvars):
```
Comment (jgsogo) [inline conans/client/build/cmake.py]: Maybe the `post=self._append_vcvars` makes sense as an argument to `with tools.vcvars(..., post=self._append_vcvars)` and keep `environment_append` unmodified.

**Comment by memsharded** [comment]
Hi @agurtovoy 

Please check the description:

> Changelog: Feature: Adds vcvars_append variable (defaulting to False) to CMake and Meson build helpers constructors, so when they need to activate the Visual Studio environment via vcvars (for Ninja and NMake generators), the vcvars environment is appended at the end, giving precedence to the environment previously defined.

As the risk of breaking was still there, and there were discussions if the build helper should actually have precedence, we have considered it an opt-in feature at the moment. Will probably change the default in Conan 2.0, but at the moment, better not risk.

**Comment by agurtovoy** [comment]
Hi @memsharded, the description looks good to me. Thanks again for yours and your team's work on this.

**Pull request merged**